### DISCUSSION TITLE
How to install/enable FIPS from compiled from one machine to use in another machine?

### DISCUSSION THREAD

**Discussion by renildk7878** [openssl/openssl#25373] [Q&A]
Please do NOT use issues to ask questions about OpenSSL.

Instead, please use the [Q&A category in Discussions](<https://github.com/openssl/openssl/discussions/new?category=q-a>)
to ask your question.

Hi team,

I have question, we have requirement where we need to be moving towards FIPS compliant in out software. Many micro services are golang based. We have started to use microsoft's golang fork to support FIPS. We are building the binaries with systemrcypto; Upon running it need to have openssl + FIPS on host. The nature of our software won't allow us to build openssl during installation to support golden image concept. What we are trying to achieve is build openssl in dev/devop env with FIPS using this link: https://github.com/openssl/openssl/blob/master/README-FIPS.md#download-and-build-a-validated-fips-provider with PREFIX Tag. After that we are creating tar ball of PREFIX dir. Untar it in where our installer runs. We do set Correct PATH and LD_LIBRARY_PATH from file location of extraction of tarball. Then try to run binary but it throws this error:  panic: opensslcrypto: can't enable FIPS mode for OpenSSL 3.1.6 4 Jun 2024: OSSL_PROVIDER_try_load
openssl error(s):
error:12800067:DSO support routines::could not load the shared library
        crypto/dso/dso_dlfcn.c:118
error:12800067:DSO support routines::could not load the shared library
        crypto/dso/dso_lib.c:152
error:07880025:common libcrypto routines::reason(524325)
        crypto/provider_core.c:911

goroutine 1 [running]:
crypto/internal/backend.init.1()
        crypto/internal/backend/openssl_linux.go:85 +0x254

Ou goal is to compile the openssl in one host and then use it as openssl tarball. Extract it and set few things(e,g env varibles) and run the binary. Is there a correct way to use OPENSSL+FIPS compiled in one host and can be used other host??

**Comment by nhorman** [comment]
sounds like your fips provider dso isn't installed to the right location.  the default location that openssl searches for providers is:
```
${INSTALLTOP}/${LIBDIR}/ossl-modules
```
Where INSTALLTOP is the value provider during configuration via the --prefix option, and LIBDIR is either lib or lib64 on linux platforms

If you untarred your installation to a location that doesn't exactly match the path your determined with your --prefix configuration setting, this is the common result.  you can work around it by setting the OPENSSL_MODULES environment variable to the appropraite path

**Comment by renildk7878** [comment]
Does the building openssl + FIPS in one machine can work on new machine? I was sort of confused here: https://github.com/openssl/openssl/blob/master/README-FIPS.md#installing-the-fips-provider This line "The FIPS module must have the self tests run, and the FIPS module config file output generated on every machine that it is to be used on. For OpenSSL 3.0, you must not copy the FIPS module config file output data from one machine to another."

**Reply by paulidale** [reply]
Appendix A of the security policy is pretty clear.  You **must** run `openssl fipsinstall` on the target machine.  If you don't do this, you are not following the letter of the installation instructions and will not be running in an approved mode.  This doesn't mean it has to be built on the target machine, just the fipsinstall step needs to be run there.

If you are content to run in the spirit but not the letter of the FIPS validation, there is a way to avoid running fipsinstall on the target but it isn't strictly approved.

**Reply by renildk7878** [reply]
Make sense. I have followed the both the guidelines with the change. Setting up OPENSSL_MODULES as env to untar'd location and ran openssl fipsinstall. But it gives this 
![image](https://github.com/user-attachments/assets/e9d333dd-039e-40c0-bffd-7a09cb8ab995)

**Reply by paulidale** [reply]
It's not finding your configuration properly.  It needs both the config file and the fipsmodule.cnf.  You might need to set `OPENSSL_CONF_INCLUDE` to get the latter.
